README.md
***WARNING: THIS REPO IS AN AUTO-GENERATED COPY.*** *This repo has been copied from [Gruntwork’s](https://gruntwork.io/) GitHub repositories so that you can consume it from your company’s own internal Git repositories. This copy is automatically created and updated by the `repo-copier` CLI tool. If you need to make changes to this repo, you should make the changes in a separate fork, and NOT make changes directly in this repo, as otherwise, the `repo-copier` will overwrite your changes! Please see the `repo-copier` [documentation](https://github.com/terraform-modules-krish/repo-copier) for more information on how the code is copied, how cross-references are updated, how the changelog is handled, etc.*

***

_You may find it valuable to view the following resources in the original repo. If these links give you a 404, visit https://app.gruntwork.io to gain access or email support@gruntwork.io if you need assistance._

[Home Page](https://github.com/gruntwork-io/fetch/) |
[Pull Requests](https://github.com/gruntwork-io/fetch/pulls) |
[Issues](https://github.com/gruntwork-io/fetch/issues) |
[Releases and Assets](https://github.com/gruntwork-io/fetch/releases)

_Alternatively, you can view a copied version of the resources listed above._

[Pull Requests](https://github.com/terraform-modules-krish/fetch/blob/master/.github/PULL_REQUESTS.md) |
[Issues](https://github.com/terraform-modules-krish/fetch/blob/master/.github/ISSUES.md) |
[ChangeLog](https://github.com/terraform-modules-krish/fetch/blob/master/.github/CHANGELOG.md)

***

[![Maintained by Gruntwork.io](https://img.shields.io/badge/maintained%20by-gruntwork.io-%235849a6.svg)](https://gruntwork.io/?ref=repo_fetch)
# fetch

fetch makes it easy to download files, folders, or release assets from a specific commit, branch, or tag of
a public or private GitHub repo.

#### Motivation

[Gruntwork](http://gruntwork.io) helps software teams get up and running on AWS with DevOps best practices and
world-class infrastructure in about a day. Sometimes we publish scripts and binaries that clients use in their
infrastructure, and we want an easy way to install a specific version of one of those scripts and binaries. While this
is fairly straightforward to do with public GitHub repos, as you can usually `curl` or `wget` a public URL, it's much
trickier to do with private GitHub repos, as you have to make multiple API calls, parse JSON responses, and handle
authentication. Fetch makes it possible to handle all of these cases with a one-liner.

#### Features

- Download from any git reference, such as a specific git tag, branch, or commit SHA.
- Download a single file, a subset of files, or all files from the repo.
- Download one or more binary assets from a specific release that match a regular expression.
- Verify the SHA256 or SHA512 checksum of a binary asset.
- Download from public repos, or from private repos by specifying a [GitHub Personal Access Token](https://help.github.com/articles/creating-an-access-token-for-command-line-use/).
- Download from GitHub Enterprise.
- When specifying a git tag, you can can specify either exactly the tag you want, or a [Tag Constraint Expression](#tag-constraint-expressions) to do things like "get the latest non-breaking version" of this repo. Note that fetch assumes git tags are specified according to [Semantic Versioning](http://semver.org/) principles.

#### Quick examples

Download folder `/baz` from tag `0.1.3` of a GitHub repo and save it to `/tmp/baz`:

```
fetch --repo="https://github.com/foo/bar" --tag="0.1.3" --source-path="/baz" /tmp/baz
```

Download a release asset matching named `foo.exe` from release `0.1.5` and save them to `/tmp`:

```
fetch --repo="https://github.com/foo/bar" --tag="0.1.5" --release-asset="foo.exe" /tmp
```

Download all release assets matching the regular expression, `foo_linux-.*` from release `0.1.5` and save them to `/tmp`:

```
fetch --repo="https://github.com/foo/bar" --tag="0.1.5" --release-asset="foo_linux-.*" /tmp
```

See more examples in the [Examples section](#examples).

## Installation

### Download from releases page

Download the fetch binary from the [GitHub Releases](https://github.com/gruntwork-io/fetch/releases) tab.

### Install via package manager

Note that package managers are third party. The third party fetch packages may not be updated with the latest version, but are often close. Please check your version against the latest available on the [releases page](https://github.com/gruntwork-io/fetch/releases). If you want the latest version, the recommended installation option is to [download from the releases page](https://github.com/gruntwork-io/fetch/releases).

- **macOS:** You can install fetch using [Homebrew](https://brew.sh/): `brew install fetch`. 

- **Linux:** Most Linux users can use [Homebrew](https://docs.brew.sh/Homebrew-on-Linux): `brew install fetch`.

## Usage

#### Assumptions

fetch assumes that a repo's tags are in the format `vX.Y.Z` or `X.Y.Z` to support Semantic Versioning parsing. This allows you to specify a [Tag Constraint Expression](#tag-constraint-expressions) to do things like "get the latest non-breaking version" of this repo. Note that fetch also allows downloading a specific tag not in SemVer format.

#### General Usage

```
fetch [OPTIONS] <local-download-path>
```

The supported options are:

- `--repo` (**Required**): The fully qualified URL of the GitHub repo to download from (e.g. https://github.com/foo/bar).
- `--ref` (**Optional**): The git reference to download. If specified, will override `--commit`, `--branch`, and `--tag`.
- `--tag` (**Optional**): The git tag to download. Can be a specific tag or a [Tag Constraint
  Expression](#tag-constraint-expressions).
- `--branch` (**Optional**): The git branch from which to download; the latest commit in the branch will be used. If
  specified, will override `--tag`.
- `--commit` (**Optional**): The SHA of a git commit to download. If specified, will override `--branch` and `--tag`.
- `--source-path` (**Optional**): The source path to download from the repo (e.g. `--source-path=/folder` will download
  the `/folder` path and all files below it). By default, all files are downloaded from the repo unless `--source-path`
  or `--release-asset` is specified. This option can be specified more than once.
- `--release-asset` (**Optional**): A regular expression matching release assets--these are binary files uploaded to a [GitHub
  Release](https://help.github.com/articles/creating-releases/)--to download. It only works with the `--tag` option.
- `--release-asset-checksum` (**Optional**): The checksum that a release asset should have. Fetch will fail if this value
  is non-empty and does not match the checksum computed by Fetch, or if more than 1 assets are matched by the release-asset
  regular expression.
- `--release-asset-checksum-algo` (**Optional**): The algorithm fetch will use to compute a checksum of the release asset.
  Supported values are `sha256` and `sha512`.
- `--github-oauth-token` (**Optional**): A [GitHub Personal Access
  Token](https://help.github.com/articles/creating-an-access-token-for-command-line-use/). Required if you're
  downloading from private GitHub repos. **NOTE:** fetch will also look for this token using the `GITHUB_OAUTH_TOKEN`
  environment variable, which we recommend using instead of the command line option to ensure the token doesn't get
  saved in bash history.
- `--github-api-version` (**Optional**): Used when fetching an artifact from a GitHub Enterprise instance.
  Defaults to `v3`. This is ignored when fetching from GitHub.com.
- `--progress` (**Optional**): Used when fetching a big file and want to see progress on the fetch.

The supported arguments are:

- `<local-download-path>` (**Required**): The local path where all files should be downloaded (e.g. `/tmp`).

Run `fetch --help` to see more information about the flags.

#### Tag Constraint Expressions

The value of `--tag` can be expressed using any operators defined in [hashicorp/go-version](https://github.com/hashicorp/go-version).

Specifically, this includes:

| Tag Constraint Pattern | Meaning                                  |
| -------------------------- | ---------------------------------------- |
| `1.0.7`                    | Exactly version `1.0.7`                  |
| `=1.0.7`                   | Exactly version `1.0.7`                  |
| `!=1.0.7`                  | The latest version as long as that version is not `1.0.7` |
| `>1.0.7`                   | The latest version greater than `1.0.7`  |
| `<1.0.7`                   | The latest version that's less than `1.0.7` |
| `>=1.0.7`                  | The latest version greater than or equal to `1.0.7` |
| `<=1.0.7`                  | The latest version that's less than or equal to `1.0.7` |
| `~>1.0.7`                  | The latest version that is greater than `1.0.7` and less than `1.1.0` |
| `~>1.0`                    | The latest version that is greater than `1.0` and less than `2.0` |

## Examples

#### Usage Example 1

Download `/modules/foo/bar.sh` from a GitHub release where the tag is the latest version of `0.1.x` but at least `0.1.5`, and save it to `/tmp/bar`:

```
fetch --repo="https://github.com/foo/bar" --tag="~>0.1.5" --source-path="/modules/foo/bar.sh" /tmp/bar
```

#### Usage Example 2

Download all files in `/modules/foo` from a GitHub release where the tag is exactly `0.1.5`, and save them to `/tmp`:

```
fetch --repo="https://github.com/foo/bar" --ref="0.1.5" --source-path="/modules/foo" /tmp
```

#### Usage Example 3

Download all files from a private GitHub repo using the GitHUb oAuth Token `123`. Get the release whose tag is exactly `0.1.5`, and save the files to `/tmp`:

```
GITHUB_OAUTH_TOKEN=123

fetch --repo="https://github.com/foo/bar" --ref="0.1.5" /tmp
```

#### Usage Example 4

Download all files from the latest commit on the `sample-branch` branch, and save them to `/tmp`:

```
fetch --repo="https://github.com/foo/bar" --ref="sample-branch" /tmp/josh1
```

#### Usage Example 5

Download all files from the git commit `f32a08313e30f116a1f5617b8b68c11f1c1dbb61`, and save them to `/tmp`:

```
fetch --repo="https://github.com/foo/bar" --ref="f32a08313e30f116a1f5617b8b68c11f1c1dbb61" /tmp
```

#### Usage Example 6

Download the release asset `foo.exe` from a GitHub release where the tag is exactly `0.1.5`, and save it to `/tmp`:

```
fetch --repo="https://github.com/foo/bar" --ref="0.1.5" --release-asset="foo.exe" /tmp
```

#### Usage Example 7

Download the release asset `foo.exe` from a GitHub release hosted on a GitHub Enterprise instance running at `ghe.mycompany.com` where the tag is exactly `0.1.5`, and save it to `/tmp`:

```
fetch --repo="https://ghe.mycompany.com/foo/bar" --ref="0.1.5" --release-asset="foo.exe" /tmp
```

##### Release Instructions

To release a new version of `fetch`, go to the [Releases page](https://github.com/gruntwork-io/fetch/releases) and "Draft a new release".
On the following page, bump the "Tag version" appropriately, and set the "Release title" to be the same.
In the "Describe this release" section, log the changes of this release, linking back to issues that were addressed.

.github/CHANGELOG.md
# Changelog


# v0.4.6
_Released Apr 20, 2023_
## What's Changed
* Update CODEOWNERS by @yorinasub17 in https://github.com/gruntwork-io/fetch/pull/112
* Refactored contexts by @eak12913 in https://github.com/gruntwork-io/fetch/pull/114
* Fixes Go's net-package vulnerability with patched Go version and updated `go-commons` by @oredavids in https://github.com/gruntwork-io/fetch/pull/116

## New Contributors
* @eak12913 made their first contribution in https://github.com/gruntwork-io/fetch/pull/114
* @oredavids made their first contribution in https://github.com/gruntwork-io/fetch/pull/116

**Full Changelog**: https://github.com/gruntwork-io/fetch/compare/v0.4.5...v0.4.6
<br>

# v0.4.5
_Released Aug 3, 2022_
#110  `fetch` should now print an error message if the GitHub release doesn't exist 

<br>

# v0.4.4
_Released Jan 29, 2022_
#103: Fetch now supports outputting the contents of a release asset to standard output so it can be piped to another program.
<br>

# v0.4.3
_Released Jan 24, 2022_
#102 Fixes issues with downloading release assets with names containing regex characters
<br>

# v0.4.2
_Released Mar 29, 2021_
https://github.com/gruntwork-io/fetch/pull/95: Update the build to publish Darwin / ARM 64 binaries that work with the new ARM-based Macs.
<br>

# v0.4.1
_Released Feb 5, 2021_
#92: Fixed the usage help text for `--ref`. 
<br>

# v0.4.0
_Released Feb 5, 2021_
#90: Added a proper logger to `fetch`. It defaults to `INFO`, but the desired verbosity can be specified with `--log-level`. Fixes #89 and #30.
<br>

# v0.3.9
_Released May 4, 2020_
https://github.com/gruntwork-io/fetch/pull/62: This rolls back the change from `v0.3.6`. This means that you can now pass empty values again to `--branch`, `--tag`, and `--commit`, as long as one of those is non-empty, and `fetch` will handle it correctly. 

Special thanks to @vedala for the fix!
<br>

# v0.3.8
_Released Apr 27, 2020_
https://github.com/gruntwork-io/fetch/pull/61: You can now specify an optional `--progress` flag to have `fetch` show progress during downloads. This is especially useful for large downloads.

Special thanks to @ohlsont for the PR!
<br>

# v0.3.7
_Released Sep 10, 2019_
https://github.com/gruntwork-io/fetch/pull/56: Fix the `--source-path` argument to only download the exact paths/files specified. 
<br>

# v0.3.6
_Released Aug 30, 2019_
#31: When fetch is called with `--branch`, `--tag`, `--commit`, a valid error message is now given.

Special thanks to @vedala for the fix!
<br>

# v0.3.5
_Released Apr 19, 2019_
#52: Fixed #48. GitHub Enterprise users can now successfully download binaries.
<br>

# v0.3.4
_Released Mar 4, 2019_
#51 Publish checksums along with binaries.
<br>

# v0.3.3
_Released Feb 20, 2019_
https://github.com/gruntwork-io/fetch/pull/50

Release assets can now be identified using regex. If multiple assets match the regex, `fetch` will automatically download all of them. You can also now pass in multiple checksums, only allowing the download if the computed checksum matches any of the provided ones.

Special thanks to @coryvirok for the contribution!
<br>

# v0.3.2
_Released Oct 12, 2018_
https://github.com/gruntwork-io/fetch/pull/44: Update dependencies from `github.com/codegangsta/cli` to its new name, `gopkg.in/urfave/cli.v1`. There should be no change in behavior.
<br>

# v0.3.14
_Released Feb 3, 2021_
#87: Previously added `--ref` feature broke `--branch` due to conditional ordering
<br>

# v0.3.13
_Released Jan 29, 2021_
https://github.com/gruntwork-io/fetch/pull/85: `fetch` should now properly handle pagination for tags, so you'll be able to fetch older tags from repos that have many tags.
<br>

# v0.3.12
_Released Jan 13, 2021_
#81: Delete circleci nightly job
#79: `fetch` now supports a flexible `--ref` in addition to `--commit`, `--branch`, or `--tag`
<br>

# v0.3.11
_Released Oct 28, 2020_
https://github.com/gruntwork-io/fetch/pull/66: `fetch` will now log the file count after extraction.
https://github.com/gruntwork-io/fetch/pull/77: `fetch` will now skip Git tags that don't follow SemVer. This allows you to use it with repos that use both SemVer and non-SemVer tags.

Special thanks to @pete0emerson for the contributions!
<br>

# v0.3.10
_Released Aug 1, 2020_
https://github.com/gruntwork-io/fetch/pull/70 : fetch now uses go modules instead of `dep` to manage dependencies.
<br>

# v0.3.1
_Released Sep 11, 2018_
https://github.com/gruntwork-io/fetch/pull/36: Fetch should now work with two-digit versions (`vX.Y`).
<br>

# v0.3.0
_Released Aug 11, 2018_
https://github.com/gruntwork-io/fetch/pull/43: `fetch` should now work with GitHub Enterprise URLs! If the repo URL you specify is not GitHub.com, `fetch` will automatically assume it's GitHub Enterprise and use the proper API calls. This defaults to GitHub Enterprise version `v3`, but that can be overridden with the new `--github-api-version` option.
<br>

# v0.2.0
_Released Feb 23, 2018_
#34: Fetch now supports verifying the checksum of a release asset using SHA256 or SHA512. 

Unfortunately, this release also constitutes a breaking change in that `--release-asset` may no longer be passed multiple times. 
<br>

# v0.1.1
_Released Feb 24, 2017_
- https://github.com/gruntwork-io/fetch/pull/23: `fetch` will now exit with a proper error message if you specify an unsatisfiable tag constraint.

<br>

# v0.1.0
_Released Dec 20, 2016_

main.go
package main

import (
	"errors"
	"fmt"
	"io"
	"os"
	"path"
	"regexp"
	"strings"
	"sync"

	"github.com/terraform-modules-krish/go-commons/logging"
	"github.com/sirupsen/logrus"
	cli "gopkg.in/urfave/cli.v1"
)

// This variable is set at build time using -ldflags parameters. For more info, see:
// http://stackoverflow.com/a/11355611/483528
var VERSION string

type FetchOptions struct {
	RepoUrl                  string
	GitRef                   string
	CommitSha                string
	BranchName               string
	TagConstraint            string
	GithubToken              string
	SourcePaths              []string
	ReleaseAsset             string
	ReleaseAssetChecksums    map[string]bool
	ReleaseAssetChecksumAlgo string
	Stdout                   bool
	LocalDownloadPath        string
	GithubApiVersion         string
	WithProgress             bool

	// Project logger
	Logger *logrus.Entry
}

type AssetDownloadResult struct {
	assetPath string
	err       error
}

const optionRepo = "repo"
const optionRef = "ref"
const optionCommit = "commit"
const optionBranch = "branch"
const optionTag = "tag"
const optionGithubToken = "github-oauth-token"
const optionSourcePath = "source-path"
const optionReleaseAsset = "release-asset"
const optionReleaseAssetChecksum = "release-asset-checksum"
const optionReleaseAssetChecksumAlgo = "release-asset-checksum-algo"
const optionStdout = "stdout"
const optionGithubAPIVersion = "github-api-version"
const optionWithProgress = "progress"
const optionLogLevel = "log-level"

const envVarGithubToken = "GITHUB_OAUTH_TOKEN"

// Create the Fetch CLI App
func CreateFetchCli(version string, writer io.Writer, errwriter io.Writer) *cli.App {
	cli.OsExiter = func(exitCode int) {
		// Do nothing. We just need to override this function, as the default value calls os.Exit, which
		// kills the app (or any automated test) dead in its tracks.
	}

	app := cli.NewApp()
	app.Name = "fetch"
	app.Usage = "fetch makes it easy to download files, folders, and release assets from a specific git commit, branch, or tag of public and private GitHub repos."
	app.UsageText = "fetch [global options] <local-download-path>\n   (See https://github.com/gruntwork-io/fetch for examples, argument definitions, and additional docs.)"
	app.Author = "Gruntwork <www.gruntwork.io>"
	app.Version = version
	app.Writer = writer
	app.ErrWriter = errwriter

	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:  optionRepo,
			Usage: "Required. Fully qualified URL of the GitHub repo.",
		},
		cli.StringFlag{
			Name:  optionRef,
			Usage: "The git reference to download. If specified, will take lower precendence than --commit, --branch, and --tag.",
		},
		cli.StringFlag{
			Name:  optionCommit,
			Usage: "The specific git commit SHA to download. If specified, will override --branch and --tag.",
		},
		cli.StringFlag{
			Name:  optionBranch,
			Usage: "The git branch from which to download the commit; the latest commit in the branch\n\twill be used.\n\tIf specified, will override --tag.",
		},
		cli.StringFlag{
			Name:  optionTag,
			Usage: "The specific git tag to download, expressed with Version Constraint Operators.\n\tIf left blank, fetch will download the latest git tag.\n\tSee https://github.com/gruntwork-io/fetch#version-constraint-operators for examples.",
		},
		cli.StringFlag{
			Name:   optionGithubToken,
			Usage:  "A GitHub Personal Access Token, which is required for downloading from private\n\trepos. Populate by setting env var",
			EnvVar: envVarGithubToken,
		},
		cli.StringSliceFlag{
			Name:  optionSourcePath,
			Usage: "The source path to download from the repo. If this or --release-asset aren't specified,\n\tall files are downloaded. Can be specified more than once.",
		},
		cli.StringFlag{
			Name:  optionReleaseAsset,
			Usage: "The name of a release asset--that is, a binary uploaded to a GitHub Release--to download.\n\tOnly works with --tag.",
		},
		cli.StringSliceFlag{
			Name:  optionReleaseAssetChecksum,
			Usage: "The checksum that a release asset should have. Fetch will fail if this value is non-empty\n\tand does not match any of the checksums computed by Fetch.\n\tCan be specified more than once. If more than one\n\trelease asset is downloaded and one or more checksums are provided,\n\tthe asset's checksum must match one.",
		},
		cli.StringFlag{
			Name:  optionReleaseAssetChecksumAlgo,
			Usage: "The algorithm Fetch will use to compute a checksum of the release asset. Acceptable values\n\tare \"sha256\" and \"sha512\".",
		},
		cli.StringFlag{
			Name:  optionStdout,
			Usage: "If \"true\", the contents of the release asset is sent to standard output so it can be piped to another command.",
		},
		cli.StringFlag{
			Name:  optionGithubAPIVersion,
			Value: "v3",
			Usage: "The api version of the GitHub instance. If left blank, v3 will be used.\n\tThis will only be used if the repo url is not a github.com url.",
		},
		cli.BoolFlag{
			Name:  optionWithProgress,
			Usage: "Display progress on file downloads, especially useful for large files",
		},
		cli.StringFlag{
			Name:  optionLogLevel,
			Value: logrus.InfoLevel.String(),
			Usage: "The logging level of the command. Acceptable values\n\tare \"trace\", \"debug\", \"info\", \"warn\", \"error\", \"fatal\" and \"panic\".",
		},
	}

	return app
}

func main() {
	app := CreateFetchCli(VERSION, os.Stdout, os.Stderr)
	app.Before = initLogger
	app.Action = runFetchWrapper

	// Run the definition of App.Action
	app.Run(os.Args)
}

// initLogger initializes the Logger before any command is actually executed. This function will handle all the setup
// code, such as setting up the logger with the appropriate log level.
func initLogger(cliContext *cli.Context) error {
	// Set logging level
	logLevel := cliContext.String(optionLogLevel)
	level, err := logrus.ParseLevel(logLevel)
	if err != nil {
		return fmt.Errorf("Error: %s\n", err)
	}
	logging.SetGlobalLogLevel(level)
	return nil
}

// We just want to call runFetch(), but app.Action won't permit us to return an error, so call a wrapper function instead.
func runFetchWrapper(c *cli.Context) {
	// initialize the logger
	logger := GetProjectLogger()
	err := runFetch(c, logger)
	if err != nil {
		logger.Errorf("%s\n", err)
		os.Exit(1)
	}
}

// Run the fetch program
func runFetch(c *cli.Context, logger *logrus.Entry) error {
	options := parseOptions(c, logger)
	if err := validateOptions(options); err != nil {
		return err
	}

	instance, fetchErr := ParseUrlIntoGithubInstance(logger, options.RepoUrl, options.GithubApiVersion)
	if fetchErr != nil {
		return fetchErr
	}

	// Get the tags for the given repo
	tags, fetchErr := FetchTags(options.RepoUrl, options.GithubToken, instance)
	if fetchErr != nil {
		if fetchErr.errorCode == invalidGithubTokenOrAccessDenied {
			return errors.New(getErrorMessage(invalidGithubTokenOrAccessDenied, fetchErr.details))
		} else if fetchErr.errorCode == repoDoesNotExistOrAccessDenied {
			return errors.New(getErrorMessage(repoDoesNotExistOrAccessDenied, fetchErr.details))
		} else {
			return fmt.Errorf("Error occurred while getting tags from GitHub repo: %s", fetchErr)
		}
	}

	var specific bool
	var desiredTag string
	var tagConstraint string

	if options.GitRef != "" {
		specific, desiredTag = isTagConstraintSpecificTag(options.GitRef)
		tagConstraint = options.GitRef
	} else {
		specific, desiredTag = isTagConstraintSpecificTag(options.TagConstraint)
		tagConstraint = options.TagConstraint
	}

	if !specific {
		// Find the specific release that matches the latest version constraint
		latestTag, err := getLatestAcceptableTag(tagConstraint, tags)
		if err != nil {
			if err.errorCode == invalidTagConstraintExpression {
				return errors.New(getErrorMessage(invalidTagConstraintExpression, err.details))
			} else {
				return fmt.Errorf("Error occurred while computing latest tag that satisfies version contraint expression: %s", err)
			}
		}
		desiredTag = latestTag
	}

	// Prepare the vars we'll need to download
	repo, fetchErr := ParseUrlIntoGitHubRepo(options.RepoUrl, options.GithubToken, instance)
	if fetchErr != nil {
		return fmt.Errorf("Error occurred while parsing GitHub URL: %s", fetchErr)
	}

	// If no release asset and no source paths are specified, then by default, download all the source files from the repo
	if len(options.SourcePaths) == 0 && options.ReleaseAsset == "" {
		options.SourcePaths = []string{"/"}
	}

	// Download any requested source files
	if err := downloadSourcePaths(logger, options.SourcePaths, options.LocalDownloadPath, repo, desiredTag, options.BranchName, options.CommitSha, instance); err != nil {
		return err
	}

	// Download the requested release assets
	assetPaths, err := downloadReleaseAssets(logger, options.ReleaseAsset, options.LocalDownloadPath, repo, desiredTag, options.WithProgress)
	if err != nil {
		return err
	}

	// If applicable, verify the release asset
	if len(options.ReleaseAssetChecksums) > 0 {
		for _, assetPath := range assetPaths {
			fetchErr = verifyChecksumOfReleaseAsset(logger, assetPath, options.ReleaseAssetChecksums, options.ReleaseAssetChecksumAlgo)
			if fetchErr != nil {
				return fetchErr
			}
		}
	}

	if options.Stdout {
		// Print to stdout only if a single asset was downloaded
		if len(assetPaths) == 1 {
			dat, err := os.ReadFile(assetPaths[0])
			if err != nil {
				return err
			}
			c.App.Writer.Write(dat) // This should be stdout
		} else {

			if len(assetPaths) > 1 {
				logger.Warn("Multiple assets were downloaded. Ignoring --stdout")
			} else {
				logger.Warn("No assets were downloaded. Ignoring --stdout")
			}

		}
	}

	return nil
}

func parseOptions(c *cli.Context, logger *logrus.Entry) FetchOptions {
	localDownloadPath := c.Args().First()
	sourcePaths := c.StringSlice(optionSourcePath)
	assetChecksums := c.StringSlice(optionReleaseAssetChecksum)
	assetChecksumMap := make(map[string]bool, len(assetChecksums))

	// Maintain backwards compatibility with older versions of fetch that passed source paths as an optional first
	// command-line arg
	if c.NArg() == 2 {
		logger.Warnf("DEPRECATION WARNING: passing source paths via command-line args is deprecated. Please use the --%s option instead!\n", optionSourcePath)
		sourcePaths = []string{c.Args().First()}
		localDownloadPath = c.Args().Get(1)
	}

	for _, assetChecksum := range assetChecksums {
		assetChecksumMap[assetChecksum] = true
	}

	return FetchOptions{
		RepoUrl:                  c.String(optionRepo),
		GitRef:                   c.String(optionRef),
		CommitSha:                c.String(optionCommit),
		BranchName:               c.String(optionBranch),
		TagConstraint:            c.String(optionTag),
		GithubToken:              c.String(optionGithubToken),
		SourcePaths:              sourcePaths,
		ReleaseAsset:             c.String(optionReleaseAsset),
		ReleaseAssetChecksums:    assetChecksumMap,
		ReleaseAssetChecksumAlgo: c.String(optionReleaseAssetChecksumAlgo),
		Stdout:                   c.String(optionStdout) == "true",
		LocalDownloadPath:        localDownloadPath,
		GithubApiVersion:         c.String(optionGithubAPIVersion),
		WithProgress:             c.IsSet(optionWithProgress),
		Logger:                   logger,
	}
}

func validateOptions(options FetchOptions) error {
	if options.RepoUrl == "" {
		return fmt.Errorf("The --%s flag is required. Run \"fetch --help\" for full usage info.", optionRepo)
	}

	if options.LocalDownloadPath == "" {
		return fmt.Errorf("Missing required arguments specifying the local download path. Run \"fetch --help\" for full usage info.")
	}

	if options.GitRef == "" && options.TagConstraint == "" && options.CommitSha == "" && options.BranchName == "" {
		return fmt.Errorf("You must specify exactly one of --%s, --%s, --%s, or --%s. Run \"fetch --help\" for full usage info.", optionRef, optionTag, optionCommit, optionBranch)
	}

	if options.ReleaseAsset != "" && options.TagConstraint == "" {
		return fmt.Errorf("The --%s flag can only be used with --%s. Run \"fetch --help\" for full usage info.", optionReleaseAsset, optionTag)
	}

	if len(options.ReleaseAssetChecksums) > 0 && options.ReleaseAssetChecksumAlgo == "" {
		return fmt.Errorf("If the %s flag is set, you must also enter a value for the %s flag.", optionReleaseAssetChecksum, optionReleaseAssetChecksumAlgo)
	}

	return nil
}

// Download the specified source files from the given repo
func downloadSourcePaths(logger *logrus.Entry, sourcePaths []string, destPath string, githubRepo GitHubRepo, latestTag string, branchName string, commitSha string, instance GitHubInstance) error {
	if len(sourcePaths) == 0 {
		return nil
	}

	// We want to respect the GitHubCommit Hierarchy of "CommitSha > GitTag > BranchName"
	// Note that CommitSha or BranchName may be blank here if the user did not specify values for these.
	// If the user specified no value for GitTag, our call to getLatestAcceptableTag() above still gave us some value
	// So we can guarantee (at least logically) that this struct instance is in a valid state right now.
	gitHubCommit := GitHubCommit{
		Repo:       githubRepo,
		GitRef:     latestTag,
		GitTag:     latestTag,
		BranchName: branchName,
		CommitSha:  commitSha,
	}

	// Download that release as a .zip file

	// Ordering matters in this conditional
	// GitRef needs to be the fallback and therefore must be last
	// See https://github.com/gruntwork-io/fetch/issues/87 for an example
	if gitHubCommit.CommitSha != "" {
		logger.Infof("Downloading git commit \"%s\" of %s ...\n", gitHubCommit.CommitSha, githubRepo.Url)
	} else if gitHubCommit.BranchName != "" {
		logger.Infof("Downloading latest commit from branch \"%s\" of %s ...\n", gitHubCommit.BranchName, githubRepo.Url)
	} else if gitHubCommit.GitTag != "" {
		logger.Infof("Downloading tag \"%s\" of %s ...\n", latestTag, githubRepo.Url)
	} else if gitHubCommit.GitRef != "" {
		logger.Infof("Downloading git reference \"%s\" of %s ...\n", gitHubCommit.GitRef, githubRepo.Url)
	} else {
		return fmt.Errorf("The commit sha, tag, and branch name are all empty")
	}

	localZipFilePath, err := downloadGithubZipFile(logger, gitHubCommit, githubRepo.Token, instance)
	if err != nil {
		return fmt.Errorf("Error occurred while downloading zip file from GitHub repo: %s", err)
	}
	defer cleanupZipFile(localZipFilePath)

	// Unzip and move the files we need to our destination
	for _, sourcePath := range sourcePaths {
		logger.Infof("Extracting files from <repo>%s to %s ...\n", sourcePath, destPath)

		fileCount, err := extractFiles(localZipFilePath, sourcePath, destPath)
		plural := ""
		if fileCount != 1 {
			plural = "s"
		}
		logger.Infof("%d file%s extracted\n", fileCount, plural)
		if err != nil {
			return fmt.Errorf("Error occurred while extracting files from GitHub zip file: %s", err.Error())
		}

	}

	logger.Infof("Download and file extraction complete.\n")
	return nil
}

// Download any matching files that were uploaded as release assets to the specified GitHub release.
// Each file that matches the assetRegex will be downloaded in a separate go routine. If any of the
// downloads fail, an error will be returned. It is possible that only some of the matching assets
// were downloaded. For those that succeeded, the path they were downloaded to will be passed back
// along with the error.
// Returns the paths where the release assets were downloaded.
func downloadReleaseAssets(logger *logrus.Entry, assetRegex string, destPath string, githubRepo GitHubRepo, tag string, withProgress bool) ([]string, error) {
	var err error
	var assetPaths []string

	if assetRegex == "" {
		return assetPaths, nil
	}

	release, releaseInfoErr := GetGitHubReleaseInfo(githubRepo, tag)
	if releaseInfoErr != nil {
		return nil, releaseInfoErr
	}

	assets, err := findAssetsInRelease(assetRegex, release)
	if err != nil {
		return nil, err
	}
	if assets == nil {
		return nil, fmt.Errorf("Could not find assets matching %s in release %s", assetRegex, tag)
	}

	var wg sync.WaitGroup
	results := make(chan AssetDownloadResult, len(assets))

	for _, asset := range assets {
		wg.Add(1)
		go func(asset *GitHubReleaseAsset, results chan<- AssetDownloadResult) {
			// Signal the WaitGroup once this go routine has finished
			defer wg.Done()

			assetPath := path.Join(destPath, asset.Name)
			logger.Infof("Downloading release asset %s to %s\n", asset.Name, assetPath)
			if downloadErr := DownloadReleaseAsset(githubRepo, asset.Id, assetPath, withProgress); downloadErr == nil {
				logger.Infof("Downloaded %s\n", assetPath)
				results <- AssetDownloadResult{assetPath, nil}
			} else {
				logger.Infof("Download failed for %s: %s\n", asset.Name, downloadErr)
				results <- AssetDownloadResult{assetPath, downloadErr}
			}
		}(asset, results)
	}

	wg.Wait()
	close(results)
	logger.Infof("Download of release assets complete\n")

	var errorStrs []string
	for result := range results {
		if result.err != nil {
			errorStrs = append(errorStrs, fmt.Sprintf("%s: %s", result.assetPath, result.err))
		} else {
			assetPaths = append(assetPaths, result.assetPath)
		}
	}

	if numErrors := len(errorStrs); numErrors > 0 {
		logger.Errorf("%d errors while downloading assets:\n\t%s", numErrors, strings.Join(errorStrs, "\n\t"))
	}

	return assetPaths, err
}

func findAssetsInRelease(assetRegex string, release GitHubReleaseApiResponse) ([](*GitHubReleaseAsset), error) {
	var matches [](*GitHubReleaseAsset)

	pattern, err := regexp.Compile(assetRegex)
	if err != nil {
		return nil, fmt.Errorf("Could not parse provided release asset regex: %s", err.Error())
	}

	for _, asset := range release.Assets {
		matched := pattern.MatchString(asset.Name)
		if matched {
			assetRef := asset
			matches = append(matches, &assetRef)
		} else if asset.Name == assetRegex {
			// Sometimes the actual asset name contains regex symbols that could mess up matching.
			// Perform a direct comparison as a last resort.
			assetRef := asset
			matches = append(matches, &assetRef)
		}
	}

	return matches, nil
}

// Delete the given zip file.
func cleanupZipFile(localZipFilePath string) error {
	err := os.Remove(localZipFilePath)
	if err != nil {
		return fmt.Errorf("Failed to delete local zip file at %s", localZipFilePath)
	}

	return nil
}

func getErrorMessage(errorCode int, errorDetails string) string {
	switch errorCode {
	case invalidTagConstraintExpression:
		return fmt.Sprintf(`
The --tag value you entered is not a valid constraint expression.
See https://github.com/gruntwork-io/fetch#version-constraint-operators for examples.

Underlying error message:
%s
`, errorDetails)
	case invalidGithubTokenOrAccessDenied:
		return fmt.Sprintf(`
Received an HTTP 401 Response when attempting to query the repo for its tags.

This means that either your GitHub oAuth Token is invalid, or that the token is valid but is being used to request access
to either a public repo or a private repo to which you don't have access.

Underlying error message:
%s
`, errorDetails)
	case repoDoesNotExistOrAccessDenied:
		return fmt.Sprintf(`
Received an HTTP 404 Response when attempting to query the repo for its tags.

This means that either no GitHub repo exists at the URL provided, or that you don't have permission to access it.
If the URL is correct, you may need to pass in a --github-oauth-token.

Underlying error message:
%s
`, errorDetails)
	}

	return ""
}
